Implement path tracer

Paolo97Gll opened the pull request:
- [x] Implement function for the generation of ONB from the z-axis
- [x] Implement `scatter_ray` functions for BRDFs
- [x] Implement path tracing algorithm
- [x] Change demo to include path tracing examples
- [x] Point-light tracing and demo examples
- [x] Update documentation

Paolo97Gll committed:
Implement ONB

Co-authored-by: Samuele-Colombo <Samuele-Colombo@users.noreply.github.com>

Samuele-Colombo committed:
Rework Normal

Co-authored-by: Paolo Galli <Paolo97Gll@users.noreply.github.com>

Samuele-Colombo committed:
Add create_onb

Samuele-Colombo committed:
Add ONB test

Samuele-Colombo committed:
Merge branch 'implement_path_tracer' of https://github.com/Paolo97Gll/Raytracer.jl into implement_path_tracer

Samuele-Colombo committed:
Add needed export

Samuele-Colombo committed:
Samll tweak to create_onb

Samuele-Colombo committed:
Add PathTracer and its tests

Samuele-Colombo committed:
Improve ONB tests

Samuele-Colombo committed:
Fix ONB test

Samuele-Colombo committed:
Improve PathTracer tests (still BUGGED)

Samuele-Colombo committed:
Fix create_onb for better precision

Samuele-Colombo committed:
Improve pcg

Now it can produce an arbitrarily long sequence of Numbers (except BigInt and BigFloat) by using rand(pcg, T, n)

Paolo97Gll committed:
Update packages

Paolo97Gll committed:
Delete metaprogramming since not needed anymore

Samuele-Colombo committed:
Update CHANGELOG.md

Paolo97Gll committed:
Merge branch 'master' into implement_path_tracer

Samuele-Colombo committed:
Ensure correct type for pigment

Paolo97Gll committed:
Merge branch 'master' into implement_path_tracer

Paolo97Gll committed:
Remove type parametrization (#23)

* Change default type to Float32

* Switch to Float32

* Switch to Float32 for cameras, ray and image_tracer

* Fix typo

* Update packages

* Revert "Update packages"

This reverts commit bd46b5e7f3d069a5397b3b4a77dd91f110cdf429.

* Fix ffmpeg for Windows and convert to Float32

* Rename and split into two files

* Add const Normal with false parameter

* Fix typo

* Fix broadcasting bug

* Migrate to Float32

* Uncomment all tests

* Uncomment includes

* Fix test

* Update header

* Fix typo and add version

* Fix typo

* Add conditional multithreading to fire_all_rays!

* Change animation progress bar color to white

* Fix docstrings

Co-authored-by: Samuele-Colombo <79973069+Samuele-Colombo@users.noreply.github.com>

Samuele-Colombo committed:
Update Changelog

Paolo97Gll committed:
Add clamp to acos

Paolo97Gll committed:
Ignore ipynb

Paolo97Gll committed:
Simplify ffmpeg command

Paolo97Gll committed:
Fix arg_type of delta_theta

Paolo97Gll committed:
Fix default of delta_theta

Paolo97Gll committed:
Minor fix

Paolo97Gll committed:
Change parameter type of demoanimationloop

Paolo97Gll committed:
Update usage informations

Paolo97Gll committed:
Update license

Paolo97Gll committed:
Add custom colors

Paolo97Gll committed:
Fix typo

Paolo97Gll committed:
Migrate constructors to map since more performant

Paolo97Gll committed:
Fix typo and add path tracer

Paolo97Gll committed:
Add PCG for antialiasing and auxiliary function

Paolo97Gll committed:
Add missing default value

Paolo97Gll committed:
Fix renderer constructors

Paolo97Gll committed:
Add path tracer demo

Paolo97Gll committed:
Fix tests

Samuele-Colombo commented on src/user_utils.jl:
```diff
@@ -57,53 +112,15 @@ function demo(output_file::String,
               camera_orientation::Tuple{Float32, Float32, Float32},
               screen_distance::Float32,
-              renderer_type::Type{<:Renderer},
+              renderer_type::String,
```
I really do not like that this function argument is a string instead of a `Renderer` subtype. Using a string works well when using a CLI but it becomes cumbersome and inefficient when the code is used in the REPL or in a script. Furthermore, if we use subtypes of `Renderer` if the user makes a typo it will have a compiling error from Julia when calling the function instead of a runtime error when cheching if a `String` is valid.

Samuele-Colombo replied:
Addressed in [this commit](https://github.com/Paolo97Gll/Raytracer.jl/pull/20/commits/afc627f6f70661936287be44ff40fa415fd1a7d2)

Samuele-Colombo commented on src/user_utils.jl:
```diff
@@ -23,6 +23,61 @@ function tonemapping(input_file::String,
+                                material = Material(brdf = SpecularBRDF(pigment = UniformPigment(RGB(0.2f0, 0.6f0, 0.3f0)))))]
+        append!(world, ground, sky, other_spheres)
+        return PathTracer(world)
+    else
+        # TODO throw error
+    end
```
This is the (now missing) runtime error I was talking about in [this comment](https://github.com/Paolo97Gll/Raytracer.jl/pull/20/commits/d99df246f17e9a9fdeaaca16e180de52341c9ed0#r646335069)

Samuele-Colombo committed:
Add intervals as it is needed by AABB

Samuele-Colombo committed:
Reorganized exports to be readable and easily searchable

Samuele-Colombo committed:
Modified user utils to avoid using Strings in lower lever functions

Samuele-Colombo committed:
Fix Transformation docstrings

Samuele-Colombo committed:
Fix ordering of inclusions to avoid issues down the line

Samuele-Colombo committed:
Remove subtyping of Renderer

Samuele-Colombo committed:
Fix shapes docstrings

Samuele-Colombo committed:
Add Renderers docstrings

Samuele-Colombo committed:
Fix tests

Paolo97Gll committed:
Update default values for demo image path tracer

Paolo97Gll committed:
Merge branch 'implement_path_tracer' of https://github.com/Paolo97Gll/Raytracer.jl into implement_path_tracer

Paolo97Gll committed:
Various fix

- Change default camera_position value
- Print renderer info
- Adapt Raytracer.demo with new syntax

Paolo97Gll committed:
Various fix

- Change default camera_position value
- Print renderer info
- Adapt Raytracer.demo with new sintax

Paolo97Gll committed:
Revert "Various fix"

This reverts commit 2626cb690e4277b24116a845969586ab1c0145b1.

Paolo97Gll committed:
Various fix

- Demo is now only keyword-based with default values
- Final scene for demo images
- Rename alpha and gamma with α and γ

Paolo97Gll committed:
Add force option and warning to demo image if image already exist

Paolo97Gll committed:
Update README.md

Paolo97Gll committed:
Add antialiasing

Paolo97Gll committed:
Add antialiasing and path-tracer options to demo functions

Paolo97Gll committed:
Improve error handling, add antialiasing and path-tracer options

Paolo97Gll committed:
Fix typo

Paolo97Gll committed:
fix typo

Paolo97Gll committed:
Update README.md

Paolo97Gll committed:
Update packages

Paolo97Gll committed:
Update version number

Paolo97Gll committed:
Improve docstrings

Paolo97Gll committed:
Improve docstrings

Paolo97Gll committed:
Add todo

Paolo97Gll committed:
Trim trailing whitespaces

Paolo97Gll committed:
Add new exports

Paolo97Gll committed:
Rename color.jl in colors.jl

Paolo97Gll committed:
Rename test_color.jl in test_colors.jl

Paolo97Gll committed:
Docstrings improved

Samuele-Colombo reviewed (changes requested):
Docstrings for structs should not have as their first line the constructor but only the name of the composite type itself. The docstring for the constructor should be separate!

Samuele-Colombo committed:
Fix Camera docstrings

Samuele-Colombo committed:
Fix docs typos

Samuele-Colombo committed:
Fix Camera docstrings

Samuele-Colombo committed:
Add PointLights and Lights

Samuele-Colombo committed:
Add is_point_visible function

Samuele-Colombo committed:
Implement PointLightRenderer

Samuele-Colombo committed:
Fix is_point_visible

Samuele-Colombo committed:
Add ORIGIN to test imports

Samuele-Colombo committed:
Add quick_ray_intersection testings

Samuele-Colombo committed:
Generalize normalized_dot

Samuele-Colombo committed:
Fix DiffuseBRDF evaluation

Samuele-Colombo committed:
Add PointLightTracer to demo

Samuele-Colombo committed:
Improve demo positioning for PointLightRenderer

Paolo97Gll committed:
Change name of functions

Paolo97Gll committed:
Update γ_correction parameter type

Paolo97Gll committed:
Update docstrings

Paolo97Gll committed:
Update CHANGELOG.md

Samuele-Colombo committed:
Update CHANGELOG

Paolo97Gll committed:
Fix tests

Paolo97Gll committed:
Remove unnecessary imports

Paolo97Gll committed:
Add more exports

Paolo97Gll committed:
Update docstrings

Paolo97Gll committed:
Merge branch 'implement_path_tracer' of https://github.com/Paolo97Gll/Raytracer.jl into implement_path_tracer

Paolo97Gll committed:
Add doc scripts

Paolo97Gll committed:
Render only updated docs

Paolo97Gll committed:
Update index.md

Paolo97Gll committed:
Remove doctest from examples that can cause approximation errors

Paolo97Gll committed:
Merge branch 'implement_path_tracer' into docs

Paolo97Gll committed:
Update index.md

Paolo97Gll committed:
Update .gitignore

Paolo97Gll committed:
Merge branch 'docs' into implement_path_tracer

Paolo97Gll committed:
Docstrings update

Paolo97Gll committed:
Fix docstring

Paolo97Gll committed:
Fix docstrings

Paolo97Gll committed:
Fix typo

Paolo97Gll committed:
Enable all docs

Paolo97Gll committed:
Rename files

test_hdr_image -> test_hdrimage
test_image_tracer-> test_imagetracer

Paolo97Gll committed:
Update docstrings

Paolo97Gll committed:
Add CLI doc

Paolo97Gll committed:
Fix usage

Paolo97Gll committed:
Update CLI tool usage

Paolo97Gll committed:
Update packages

Paolo97Gll committed:
Update .gitignore

Paolo97Gll committed:
Use only DiffusiveBRDF in PointLight example
